### ISSUE TITLE
[Bug]: Crash on Startup - "Object has been destroyed"

### ISSUE THREAD

**Issue by malacandrian** [FreeTubeApp/FreeTube#4762]
### Guidelines

- [X] I have encountered this bug in the [latest release of FreeTube](https://github.com/FreeTubeApp/FreeTube/releases).
- [X] I have encountered this bug in the [official downloads of FreeTube](https://github.com/FreeTubeApp/FreeTube#official-downloads).
- [X] I have searched the issue tracker for [open](https://github.com/FreeTubeApp/FreeTube/issues?q=is%3Aopen+is%3Aissue) and [closed](https://github.com/FreeTubeApp/FreeTube/issues?q=is%3Aissue+is%3Aclosed) issues that are similar to the bug report I want to file, without success.
- [X] I have searched the [documentation](https://docs.freetubeapp.io/) for information that matches the description of the bug I want to file, without success.
- [X] This issue contains only one bug.

### Describe the bug

1. Install Freetube
2. Open Freetube
3. Close Freetube
4. Repeat 2-3 until crash happens

![image](https://github.com/FreeTubeApp/FreeTube/assets/858304/144c3f00-7a9f-4ca4-adf5-cc1c8afe5326)

> Uncaught Exception:
> TypeError: Object has been destroyed at App.<anonymous> (%AppData%\Local\Programs\FreeTube\resources\app.asar\di...:242736) at App.emit (node:events:513:28)

Once this crash has occurred once it will happen every time until Freetube is uninstalled & reinstalled.

### Expected Behavior

Freetube should open as normal

### Issue Labels

causes crash

### FreeTube Version

v0.19.2

### Operating System Version

Windows 10 Pro 22H2 19045.4170

### Installation Method

.exe

### Primary API used

Local API

### Additional Information

I have been experiencing this error sporadically for at least 3 months, possibly closer to 6. So it has persisted across multiple different versions of FreeTube.

While it is also a Javascript error crash-on-start, is a different issue to #2131 because the error is different - this one is "Object has been destroyed".

It could be related to #1250 as the error is the same, but the referenced files and lines appear to be different.

### Nightly Build

- [ ] I have encountered this bug in the latest [nightly build](https://docs.freetubeapp.io/development/nightly-builds).

**Comment by vF360** [comment]
I'm experiencing the same problem too. I guess it is easier to code a script that auto uninstalls & reinstalls at this point rather than waiting for a fix.

**Comment by hboetes** [comment]
TL;DR freetube is still running in the background, you will have to kill it.

**Comment by ngoomie** [comment]
Had this happen today with the latest Flatpak release of FreeTube, can confirm that just killing it fixed it. Very odd either way though.

**Comment by ArieKindicator** [comment]
I have the same error message. FreeTube opens with a black screen.

Rebooting or reinstalling FT didn't solve the issue.

**Comment by ArieKindicator** [comment]
Follow up: I've used FreeTube for several days now (from the date of my previous post above) without issues and "Autoplay Videos" disabled.
Today, after I (re)enabled "Autoplay Videos" again, the application crashed (randomly) again. when clicking a video to watch.

It seems that after a crash (some of) the db files are corrupt, and the app requires a fresh installation and importing the YT subscription.csv again.

Is is possible to auto backup critical db files? To be used to restore FreeTube after a crash?

**Comment by srolandmarshall** [comment]
> I'm getting this error too on macOS. I haven't discovered a clear pattern as to when it happens, but I trigger it frequently by having LibRedirect send all YouTube URLs to FreeTube, at which point as soon as it tries to open the URL it tosses up a dialog box with the error. However now the app is running in the background, and if I keep sending the URL to FreeTube I keep getting the error over and over. But then if I switch to the app (i.e. make it frontmost), it opens a window, at which point sending it a URL will work.
> 
> The specific error I'm getting is:
> 
> ```
> A JavaScript error occurred in the main process
> 
> Uncaught Exception:
> TypeError: Object has been destroyed
> at App.<anonymous> (/Applications/FreeTube.app/Contents/Resources/app.asar/dist/main.js:2:252266)
> at App.emit (node:events:514:28)```
> ```

@briankendall I am also experiencing this with the macOS version. 

my redirector just runs `open -a [FREETUBE_APP_PATH] [URL_STRING]` behind the scenes. When I try this in terminal, `open -a FREETUBE` **without the URL** works when a FreeTube window is open or closed. With the URL string, it does not work unless there is an open window. To me, this means that FreeTube may not be exiting fully when you close a window, but without digging in it's hard to determine.

I am going to take a look shortly and see if there's anything glaring.

**Comment by brian-kendall-tb** [comment]
@srolandmarshall That's the same pattern I'm seeing. Any time my browser tries to redirect a URL to FreeTube and FreeTube is already running with no open windows, it triggers that error. I can work around it by quitting FreeTube whenever I close its final window.

This may not be the same error as what the OP reported, given that it's likely mac specific. But it looks like FreeTube either needs to always quit after closing its last window, or open a new window when responding to URLs.

**Comment by briankendall** [comment]
@srolandmarshall That's the same pattern I'm seeing. Any time my browser tries to redirect a URL to FreeTube and FreeTube is already running with no open windows, it triggers that error. I can work around it by quitting FreeTube whenever I close its final window.

This may not be the same error as what the OP reported, given that it's likely mac specific. But it looks like FreeTube either needs to always quit after closing its last window, or open a new window when responding to URLs.

**Comment by skyrl69** [comment]
The bug still happens with recent v21.
I'm in a Win10 x64 environnement, the bug appeared in april with v0.20 beta. 
Not sure I downloaded 19.2 beta of 24th feb though

I used to to use PrivacyRedirects Chromium ext. Now I use LibRedirect.

It is systematicaly reproductible : Each time I close Freetube, whatever the way I opened it, whatever the way I close it, it stays in memory says here Process Explorer :
![image](https://github.com/FreeTubeApp/FreeTube/assets/45392236/89e0e048-5474-4732-84d6-764edc0e9834)

Then, when I try to restart it, whatever I have that message box
![image](https://github.com/FreeTubeApp/FreeTube/assets/45392236/5c8c1ad5-c1d9-463b-914e-f6284c13f07d)

I have to manualy kill task in process explorer to be able to run freetube

**Comment by iCaran** [comment]
I'm not 100% sure, but maybe this may offer some light:

https://github.com/FreeTubeApp/FreeTube/issues/5553#issuecomment-2286996951

**Comment by klonoa-wahoo** [comment]
This worked for me.
Try going into your 'Task Manager' on Windows
You should see 1 or more tabs of FreeTube on there
Make sure to 'End Task' them
Then try opening the app again

**Comment by LinuxOnTheDesktop** [comment]
Whether or not this error is attributable to FreeTube or to libdirect (and I, like at least one poster here, use both), there is a secondary problem that seems very much to be FreeTube's. That problem is as follows.

FreeTube does not display an informative error message when it has an instances problem (i.e. when one launches FreeTube but there is another - somehow orphaned or hung? - FreeTube process.)

How about fixing _that_ problem? :)

**Comment by srbetasign** [comment]
Hello,

i was facing the same issue.

And the solution in https://github.com/FreeTubeApp/FreeTube/issues/5553#issuecomment-2286996951 helps me to fix it.

Thank you.

**Comment by srolandmarshall** [comment]
This is still an issue on v0.22 btw.

**Comment by Kazcade** [comment]
I am also (newly) encountering this problem on 0.22, while this was not an issue with prior version. The application (or rather the -gpu and -utility subprocsses) do not terminate promptly with the renderer but stay open for ~5-10 seconds (sometimes longer) after freetube is exited.

**Comment by srolandmarshall** [comment]
I can't even launch the app directly at this point, unfortunately.

**Comment by Honytawk** [comment]
> This is still an issue on v0.22 btw. I checked out `profiles.db` and saw no such "corruption" and even deleted the extra profile that I no longer needed.

See if your profiles.db has a bunch of white spaces and tabs at the end and delete all of them.
It doesn't look corrupted, but it is.
The final characters in that file should be "}]}", everything after is corruption.

Removing those white spaces fixed the issue for me, could open FreeTube again and everything loaded as expected.

**Comment by davidolmos** [comment]
Iface the same issue since today (was running fine before), when I try to open Freetube in a Raspberry Pi 5 (8GB RAM) running Raspberry Pi OS (latest version, updated), Freetube (freetube_0.22.0_arm64.deb) does not open and I get the following error in a pop-up window:

"Uncaught Exception
TypeError: Object has been destroyed
at App <anonymous> (/opt/Freetube/resources/app.asar/distr/main.js:
2:99096)
  at App.emit (node:events:519:28)"

**Comment by iCaran** [comment]
> I face the same issue since today (was running fine before), when I try to open Freetube in a Raspberry Pi 4 (8GB RAM) running Raspberry Pi OS (latest version, updated), Freetube (freetube_0.22.0_arm64.deb) does not open and I get the following error in a pop-up window:
> 
> "Uncaught Exception TypeError: Object has been destroyed at App (/opt/Freetube/resources/app.asar/distr/main.js: 2:99096) at App.emit (node:events:519:28)"

Did you try this:- https://github.com/FreeTubeApp/FreeTube/issues/4762#issuecomment-2371003292

**Comment by davidolmos** [comment]
> > I face the same issue since today (was running fine before), when I try to open Freetube in a Raspberry Pi 4 (8GB RAM) running Raspberry Pi OS (latest version, updated), Freetube (freetube_0.22.0_arm64.deb) does not open and I get the following error in a pop-up window:
> > "Uncaught Exception TypeError: Object has been destroyed at App (/opt/Freetube/resources/app.asar/distr/main.js: 2:99096) at App.emit (node:events:519:28)"
> 
> Did you try this:- [#4762 (comment)](https://github.com/FreeTubeApp/FreeTube/issues/4762#issuecomment-2371003292)

I have fixed it by uninstalling, deleting the folders in $HOME/.config/Freetube and then reinstalling via PiApps. As Ihad all backed up, I rolled back to what I had just few days back.

Thanks for having answered so fast.

PS: It may be the solution you hinted at could have worked, but did not try it.

**Comment by srolandmarshall** [comment]
The profiles.db issue is not the same issue I am experiencing. Mine is seemingly an Electron/javascript issue in not ending processes fully and thus not being able to start new processes safely.

**Comment by iCaran** [comment]
> https://github.com/FreeTubeApp/FreeTube/issues/4762#issuecomment-2468829086

This makes sense @absidue

**Comment by iCaran** [comment]
> https://github.com/FreeTubeApp/FreeTube/issues/4762#issuecomment-2468829086

This makes sense @absidue

**Comment by srolandmarshall** [comment]
@malacandrian 
Fair, if your original issue was solved, I can make a new ticket regarding my issue. I worry that some of the litigation here that shows it's not any of the previous `Object has been destroyed` issue posts, but alas.

**Comment by MStankiewiczOfficial** [comment]
> The profiles.db issue is not the same issue I am experiencing. Mine is seemingly an Electron/javascript issue in not ending processes fully and thus not being able to start new processes safely. I have completely removed my subscriptions and profiles from FreeTube to remove the possibility of such interference and still encounter this issue only in the following scenarios:
> 
>     * I open a video.
> 
>     * I close the window.
> 
>     * I attempt to open a video via a redirect (using `freetube://` or similar `open` terminal command using `freetube`
> 
>     * I get the following error:
> 
> 
> ```
> Uncaught Exception:
> TypeError: Object has been destroyed
> at App.<anonymous> (/Applications/FreeTube.app/Contents/Resources/app.asar/dist/main.js:2:111173)
> at App.emit (node:events:519:28)
> ```
> 
> <img alt="Screen Shot 2024-11-08 at 7 47 43 AM" width="372" src="https://private-user-images.githubusercontent.com/12804079/384374479-df4c0863-753f-421e-93a8-2aa7ba0e8925.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzQ1MzAxMjksIm5iZiI6MTczNDUyOTgyOSwicGF0aCI6Ii8xMjgwNDA3OS8zODQzNzQ0NzktZGY0YzA4NjMtNzUzZi00MjFlLTkzYTgtMmFhN2JhMGU4OTI1LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDEyMTglMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMjE4VDEzNTAyOVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWYwZTk3OTgxYjMyZGM3ZGEwOWEwNGY5Y2EwOGFjNjFhOTM1NzA4ZWIxYjgxZDYzYjIyOTZkNDMyMTgwZGQ5MzAmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.rxEE-qs4XuRDGzIdzL-qxu7mjjfoPosdkjT6-HhQVNg">

This problem still exists.

**Comment by LinuxOnTheDesktop** [comment]
@Kazcade 
> If anybody is still encountering this, can you check whether disabling automatic fetching fixes the issue?

I killed the FreeTube proceses that should not be running, thereby allowing me to launch the application. Then I disabled the 'Fetch Feed Automatically' option. Then I closed the app and re-ran it. It ran without problems. But I shall have to see whether the bug recurs. (I had long had the 'fetching' enabled and only today encountered the problem for the first time.)

Perhaps then here we have a workaround (or two). What though of a fix?

**Comment by zombiemimic** [comment]
> [@Kazcade](https://github.com/Kazcade)
> 
> > If anybody is still encountering this, can you check whether disabling automatic fetching fixes the issue?
> 
> I killed the FreeTube proceses that should not be running, thereby allowing me to launch the application. Then I disabled the 'Fetch Feed Automatically' option. Then I closed the app and re-ran it. It ran without problems. But I shall have to see whether the bug recurs. (I had long had the 'fetching' enabled and only today encountered the problem for the first time.)
> 
> Perhaps then here we have a workaround (or two). What though of a fix?

I can confirm this, as of 0.23.1 Beta.

**Comment by Vyling** [comment]
Ran into the same error, but by user error. I have a few motoric problems and accidentally clicked the desktop icon four times, opening it twice. I immediately noticed that my feed didn't show the videos as already watched. I closed both windows and attempted to open FreeTube again, leading to the error.

Followed the steps above of killing remaining FreeTube processes, and it works again.

Windows 11, 24H2
FreeTube v0.23.1 Beta

**Comment by constraintAutomaton** [comment]
I also run on the same issue

Manjaro Linux 25.0.0
Freetube v0.23.2-beta

**Comment by LinuxOnTheDesktop** [comment]
Encountered on FreeTube 0.23.3 Windows 10. The problem does owe to FreeTube not shutting down properly.

**Comment by LinuxOnTheDesktop** [comment]
Bad bot.

**Comment by LinuxOnTheDesktop** [comment]
Bad bot.

**Comment by cjdg** [comment]
Happening again

**Comment by LinuxOnTheDesktop** [comment]
Bad bot.

**Comment by LinuxOnTheDesktop** [comment]
Goodness, I am sick of this bot.

**Comment by absidue** [comment]
@LinuxOnTheDesktop Rather than repeatedly complaining about the bot it would be a lot more helpful to everyone if you actually provided some steps to reproduce the issue, ideally steps that reproduce it consistently. Without any steps to reproduce the issue, there is no way anyone will be able to diagnose what is wrong and how to fix it.

**Comment by LinuxOnTheDesktop** [comment]
@absidue: you are the first to ask me that. Indeed no-one has pressed anyone in this thread for details. Nor can I supply any. For, I have not seen the problem for a while. Possibly the issue is indeed a thing of the past; I had discounted the bot's suggestion to that effect - see the story of the boy who cried wolf.

**Comment by absidue** [comment]
In that case we'll give it one more week, if nobody replies with steps to reproduce the issue on 0.23.7 or the lastest nightly build in that time, we will close this issue.

**Comment by malacandrian** [comment]
For what it's worth I've not seen this recurring since my comment on Nov 11th 2024. I suggested closing the issue then, and am happy to close it now.

**Comment by santomet** [comment]
I have been using the app for a long time, and the error showed up only now. It shows up for both the Debian package and AppImage. I am running Linux Mint 22. 

<img width="606" height="218" alt="Image" src="https://github.com/user-attachments/assets/1f05f823-2d6f-45d1-a602-5001ec5d5992" />

Interestingly, I have tried all 0.23.x versions (even those that worked well before), so apparently, there is probably something wrong with the system. Maybe corrupt js packages?

@tm2k7, did you manage to make it work?

This is what shows up when running the latest version with "--trace-warnings":

```
(node:811442) UnhandledPromiseRejectionWarning: Error: 18% of the data file is corrupt, more than given corruptAlertThreshold (10%). Cautiously refusing to start NeDB to prevent dataloss.
    at u.<anonymous> (/opt/FreeTube/resources/app.asar/dist/main.js:2:38857)
    at u.emit (node:events:518:28)
    at endReadableNT (node:internal/streams/readable:1698:12)
    at process.processTicksAndRejections (node:internal/process/task_queues:82:21)
    at emitUnhandledRejectionWarning (node:internal/process/promises:298:15)
    at warnWithErrorCodeUnhandledRejectionsMode (node:internal/process/promises:406:5)
    at processPromiseRejections (node:internal/process/promises:470:17)
    at process.processTicksAndRejections (node:internal/process/task_queues:96:32)
(node:811442) Error: 18% of the data file is corrupt, more than given corruptAlertThreshold (10%). Cautiously refusing to start NeDB to prevent dataloss.
    at u.<anonymous> (/opt/FreeTube/resources/app.asar/dist/main.js:2:38857)
    at u.emit (node:events:518:28)
    at endReadableNT (node:internal/streams/readable:1698:12)
    at process.processTicksAndRejections (node:internal/process/task_queues:82:21)


```

**Comment by aelcho19** [comment]
in my case on fedora, the problem seems to have been related to one of the .db files, when i deleted the history.db, it started working again as intended (shuts down properly). And if i re-import my history, it would revert to the same problem, luckily losing history is not a big problem and it properly stores new history, so something might have corrupted the file while upgrading versions

**Comment by constraintAutomaton** [comment]
> in my case on fedora, the problem seems to have been related to one of the .db files, when i deleted the history.db, it started working again as intended (shuts down properly). And if i re-import my history, it would revert to the same problem, luckily losing history is not a big problem and it properly stores new history, so something might have corrupted the file while upgrading versions

Where `history.db` is located?

**Comment by aelcho19** [comment]
> > in my case on fedora, the problem seems to have been related to one of the .db files, when i deleted the history.db, it started working again as intended (shuts down properly). And if i re-import my history, it would revert to the same problem, luckily losing history is not a big problem and it properly stores new history, so something might have corrupted the file while upgrading versions
> 
> Where `history.db` is located?

for me on linux it was in ~/.config/FreeTube/
if you're on windows its probably C:\Users\<YourUsername>\AppData\Roaming\FreeTube\

**Comment by constraintAutomaton** [comment]
> in my case on fedora, the problem seems to have been related to one of the .db files, when i deleted the history.db, it started working again as intended (shuts down properly). And if i re-import my history, it would revert to the same problem, luckily losing history is not a big problem and it properly stores new history, so something might have corrupted the file while upgrading versions

It did not work for me.

**Comment by protectivedad** [comment]
I tried renaming history.db, didn't work. Named it back. Restarting my computer then fixed the problem. But I had no entries in Subsciptions/VIDEOS. Closing and opening the program brought the problem back.

**Comment by haromu** [comment]
I can confirm the issue. After closing the Application, three (broken) links remain in .config/FreeTube: SingletonCookie, SingletonLock and SingletonSocket. Deleting these three solves the issue for me. Restarting the Application afterwards seems to be a bit slower though.

**Comment by pL1-JC** [comment]
Happened to me today, killed the process in task manager as instructed by @hboetes 

Worked like a charm

**Comment by Stellaeora** [comment]
> in my case on fedora, the problem seems to have been related to one of the .db files, when i deleted the history.db, it started working again as intended (shuts down properly). And if i re-import my history, it would revert to the same problem, luckily losing history is not a big problem and it properly stores new history, so something might have corrupted the file while upgrading versions

This happened to me and I was able to solve it, but it was probably a different cause.

What had happened was that a system crash (unrelated to FreeTube) while FreeTube was open had apparently caused the corruption of profiles.db. Cleaning up the corruption in the file worked for me.

<img width="1615" height="738" alt="Image" src="https://github.com/user-attachments/assets/c848d95e-10c0-49e7-a83f-8d7ee9aa85fa" />

Opening it up in a hex editor, you can see a bunch of zero bytes got appended to the file somehow, and removing those fixes it.

My advice is, if you got this error and it was due to a previous crash, clean out your cache.

**Comment by efb4f5ff-1298-471a-8973-3d47447115dc** [comment]
We have made some changes please test the latest [nightly build](https://docs.freetubeapp.io/development/nightly-builds) and let us know if has made the error go away.

**Comment by protectivedad** [comment]
> We have made some changes please test the latest [nightly build](https://docs.freetubeapp.io/development/nightly-builds) and let us know if has made the error go away.

Seems to resolve the issue for me. Previously every time I wanted to run the application twice I would have to restart my computer. That no longer seems to be the case.

**Comment by ozrendev** [comment]
> We have made some changes please test the latest [nightly build](https://docs.freetubeapp.io/development/nightly-builds) and let us know if has made the error go away.

I was also getting the error consistently (every launch had to kill process) but this seems to have fixed it for me too
